pkg/curatedpackages/reader.go

```diff
@@ -28,105 +30,117 @@ type ManifestReader interface {
+func NewPackageReader(mr ManifestReader, cache *registry.Cache) *PackageReader {
 	return &PackageReader{
 		ManifestReader: mr,
+		cache:          cache,
 	}
 }
 
 func (r *PackageReader) ReadImagesFromBundles(ctx context.Context, b *releasev1.Bundles) ([]releasev1.Image, error) {
-	images, err := r.ManifestReader.ReadImagesFromBundles(ctx, b)
-
+	var err error
+	var images []releasev1.Image
 	for _, vb := range b.Spec.VersionsBundles {
-		artifact, err := GetPackageBundleRef(vb)
+		bundleURI, bundle, err := r.getBundle(ctx, vb)
 		if err != nil {
 			logger.Info("Warning: Failed getting bundle reference", "error", err)
 			continue
 		}
-		packageImages, err := r.fetchImagesFromBundle(ctx, vb, artifact)
-		if err != nil {
-			logger.Info("Warning: Failed extracting packages", "error", err)
-			continue
-		}
+		images = append(images, releasev1.Image{URI: bundleURI})
+		packageImages := r.fetchImagesFromBundle(bundleURI, bundle)
 		images = append(images, packageImages...)
 	}
 
-	return images, err
+	return removeDuplicateImages(images), err
 }
 
 func (r *PackageReader) ReadChartsFromBundles(ctx context.Context, b *releasev1.Bundles) []releasev1.Image {
-	images := r.ManifestReader.ReadChartsFromBundles(ctx, b)
+	var images []releasev1.Image
 	for _, vb := range b.Spec.VersionsBundles {
-		artifact, err := GetPackageBundleRef(vb)
+		bundleRegistry, bundle, err := r.getBundle(ctx, vb)
 		if err != nil {
 			logger.Info("Warning: Failed getting bundle reference", "error", err)
 			continue
 		}
-		packagesHelmChart, err := fetchPackagesHelmChart(ctx, vb, artifact)
-		if err != nil {
-			logger.Info("Warning: Failed extracting packages", "error", err)
-			continue
-		}
+		images = append(images, releasev1.Image{URI: bundleRegistry})
+		packagesHelmChart := r.fetchPackagesHelmChart(bundleRegistry, bundle)
 		images = append(images, packagesHelmChart...)
 	}
-	return images
+	return removeDuplicateImages(images)
 }
 
-func fetchPackagesHelmChart(ctx context.Context, versionsBundle releasev1.VersionsBundle, artifact string) ([]releasev1.Image, error) {
-	data, err := PullLatestBundle(ctx, artifact)
+func (r *PackageReader) getBundle(ctx context.Context, vb releasev1.VersionsBundle) (string, *packagesv1.PackageBundle, error) {
+	bundleURI, err := GetPackageBundleRef(vb)
+	if err != nil {
+		return "", nil, err
+	}
+
+	artifact := registry.NewArtifactFromURI(bundleURI)
+	sc, err := r.cache.Get(registry.NewDefaultStorageContext(artifact.Registry))
+	if err != nil {
+		return "", nil, err
+	}
+
+	data, err := sc.PullBytes(ctx, artifact)
 	if err != nil {
-		return nil, err
+		return "", nil, err
 	}
-	ctrl := versionsBundle.PackageController.Controller
-	bundle := &packagesv1.PackageBundle{}
-	err = yaml.Unmarshal(data, bundle)
+	bundle := packagesv1.PackageBundle{}
+	err = yaml.Unmarshal(data, &bundle)
 	if err != nil {
-		return nil, err
+		return "", nil, err
 	}
+	return artifact.VersionedImage(), &bundle, nil
```

Comment:
What is the artifact.VersionedImage() for? Wouldn't returning just the bundle and the error suffice for this method?

Reply:
The bundle does not contain the bundles location.